### PR TITLE
[asl] Add semicolon after 'end'

### PR DESCRIPTION
**Pull request by IGGeorgiev**
As per request. This PR makes semicolons required after every statement including function bodies (`begin .. end;`)

For reference the command used for the majority of the `end` -> `end;` migration was:
```
find <target> -name '*.asl' -type f | xargs gsed -i 's/\([^A-Za-z0-9_]\|^\)end\([^A-Za-z0-9_;]\|$\)/\1end;\2/g'
```

### CODE REVIEW

**Review by Roman-Manevich — COMMENTED** [review]
Looks good to me.

**Comment by IGGeorgiev** [comment]
Updated the reference documentation as well. Let me know if I missed anything 🙂

File: asllib/doc/AbstractSyntax.tex [hunk 1]
```diff
@@ -935,4 +935,4 @@ \section{Abstract Syntax Abbreviations\label{sec:AbstractSyntaxAbreviations}}
 % \begin{itemize}
 % \item ASTRule.StmtFromList (see \secref{ASTRule.StmtFromList})
 % \item ASTRule.SequenceStmts (see \secref{ASTRule.SequenceStmts})
-% \end{itemize}
\ No newline at end of file
+% \end{itemiz}
```
Comment (Roman-Manevich) [inline asllib/doc/AbstractSyntax.tex:938]: This is commented out so it doesn't matter much, but it's better to undo this change, which, if uncommented, won't build.

File: asllib/doc/AbstractSyntax.tex [hunk 2]
```diff
@@ -692,7 +692,7 @@ \subsection{Abbreviated Rule Notation for AST Builders}
 \inferrule{}{
 {
 \begin{array}{r}
-  \buildstmt(\Nstmt(\Twhile, \punnode{\Nexpr}, \Tdo, \punnode{\Nstmtlist}, \Tend))
+  \buildstmt(\Nstmt(\Twhile, \punnode{\Nexpr}, \Tdo, \punnode{\Nstmtlist}, \Tend, \Tsemicolon))
```
Comment (Roman-Manevich) [inline asllib/doc/AbstractSyntax.tex:695]: I'm guessing you search for each occurrence of `\Tend` and added `, `\Tsemicolon` to make sure you haven't missed any, right?

Comment (IGGeorgiev) [reply asllib/doc/AbstractSyntax.tex:695]: Something like that. Some of them didn't need a comma following the `\Tend`, they needed `\parsesep` but I did double check that every occurrence of `\Tend` is followed by a semicolon in some way. (All inside the `asllib/doc` folder except for the `\Tend` definition in `asllib/doc/ASLmacros.tex`). I also did a quick search for `"end"` in the rendered pdf and it seems to have been added everywhere.

File: asllib/doc/AbstractSyntax.tex (hunk 1 above)
Comment (IGGeorgiev) [reply asllib/doc/AbstractSyntax.tex:938]: Good spot, I didn't see the missing character, I thought my editor just added a trailing newline there 😅 my bad patch incoming.

File: asllib/Parser.mly [hunk 3]
```diff
@@ -504,7 +504,7 @@ let loop_limit == { None }
 
 let stmt ==
   annotated (
-    | terminated_by(END,
+    | terminated_by(END; SEMI_COLON,
```
Comment (Roman-Manevich) [inline asllib/Parser.mly:507]: Please make this rule optional.
When an `end` is not followed by a semicolon, please emit a warning (just once for the entire run, otherwise we'll get thousands of warnings).

File: asllib/Parser.mly [hunk 4]
```diff
@@ -561,7 +561,7 @@ let params_opt == { [] } | braced(clist(opt_typed_identifier))
 let access_args == bracketed(clist(typed_identifier))
 let func_args == plist(typed_identifier)
 let maybe_empty_stmt_list == stmt_list | annotated({ S_Pass })
-let func_body == delimited(BEGIN, maybe_empty_stmt_list, END)
+let func_body == delimited(BEGIN, maybe_empty_stmt_list, END; SEMI_COLON)
```
Comment (Roman-Manevich) [inline asllib/Parser.mly:564]: Please make this rule optional.
When an `end` is not followed by a semicolon, please emit a warning (just once for the entire run, otherwise we'll get thousands of warnings).

**Review by Roman-Manevich — CHANGES_REQUESTED** [review]

File: asllib/Parser.mly [hunk 5]
```diff
@@ -90,96 +90,12 @@ let e_binop (e1, op, e2) =
...
-%nonassoc UNOPS
-
-(* IN *)
-%nonassoc IN
-
-(* DOT, brackets, etc. *)
-%left DOT LBRACKET
+%parameter<Config : sig
+    val allow_no_end_semicolon : bool
+end>
```
[the remark this answers is not in the document]
Comment (IGGeorgiev) [reply asllib/Parser.mly:98]: Sadly I don't think this is possible :/ Quoting from the menhir docs:
```
... Last, %parameter declarations take effect before %{ … %}, %token, %type, or %start declarations are considered, so that the module name uid introduced by a %parameter declaration is effectively in scope in all %{ … %}, %token, %type, or %start declarations, regardless of whether they precede or follow the %parameter declaration. This means, in particular, that the side effects of an OCaml header are observed only when the functor is applied, not when it is defined.
```

What I could do is move the parameter just above the header such that it's a bit neater.

File: asllib/builder.ml [hunk 6]
```diff
@@ -24,9 +24,10 @@ open Lexing
 
 let _debug = false
 
-type token = Parser.token
+type token = Tokens.token
 type ast_type = [ `Opn | `Ast ]
 type version = [ `ASLv0 | `ASLv1 ]
+type allow_no_end_semicolon = bool
```
[the remark this answers is not in the document]
Comment (IGGeorgiev) [reply asllib/builder.ml:30]: Yeah, I figured it might be a bit neater to alias it for when multiple bool flags appear. Happy to roll that back.

File: asllib/builder.mli [hunk 7]
```diff
@@ -24,21 +24,40 @@ open Lexing
...
+type token = Tokens.token
 type ast_type = [ `Opn | `Ast ]
 type version = [ `ASLv0 | `ASLv1 ]
+type allow_no_end_semicolon = bool
 type version_selector = [ `ASLv0 | `ASLv1 | `Any ]
 
 val from_file_result :
-  ?ast_type:ast_type -> version -> string -> AST.t Error.result
+  ?ast_type:ast_type ->
+  ?allow_no_end_semicolon:allow_no_end_semicolon ->
```
[the remark this answers is not in the document]
Comment (IGGeorgiev) [reply asllib/builder.mli:35]: The default is set to `false` in the implementation in `builder.ml`. I don't think there's a way to annotate that on the type. I could probably add a doc-string though I'm not sure what format that wants to follow?

File: asllib/Parser.mly (hunk 5 above)
[the remark this answers is not in the document]
Comment (IGGeorgiev) [reply asllib/Parser.mly:98]: Making it a `.mli` file throws an error since no `.ml` file actually exists to implement it, but `ParserConfig.ml` works.

File: asllib/builder.mli (hunk 7 above)
[the remark this answers is not in the document]
Comment (IGGeorgiev) [reply asllib/builder.mli]: This is the only comment I haven't yet addressed, do you have a preference as to where I should do this?
I'm trying to find if ocaml has a defined format for argument defaults in docstrings.

[the remark this answers is not in the document]
Comment (IGGeorgiev) [reply asllib/builder.mli]: Alright, I'll add it to the module docstring. Patch incoming in a sec.

File: asllib/error.ml [hunk 8]
```diff
@@ -232,6 +234,9 @@ module PPrint = struct
           name expected provided
     | NotYetImplemented s ->
         pp_print_text f @@ "ASL Internal error: Not yet implemented: " ^ s
+    | ObsoleteSyntax s ->
+        fprintf f "%a@ %s" pp_print_text "ASL Internal error: Obsolete syntax:"
```
[the remark this answers is not in the document]
Comment (IGGeorgiev) [reply asllib/error.ml]: fixed.

**Review by Roman-Manevich — APPROVED** [review]
Looks good to me!
